packages/marshal/src/encodePassable.js

```diff
@@ -216,11 +216,131 @@ const decodeBigInt = encoded => {
+// as `!_` and U+0021 EXCLAMATION MARK (the escape prefix) as `!|`.
+// Relative lexicographic ordering is preserved by this mapping of any character
+// at or before `!` in the contiguous range [0x00..0x21] to a respective
+// character in [0x21..0x40, 0x5F, 0x7C] preceded by `!` (which is itself in the
+// replaced range).
+// Similarly, escape `^` as `_@` and `_` as `__` because `^` indicates the start
+// of an encoded array.
+const stringEscapes = Array(0x22)
+  .fill(undefined)
+  .map((_, cp) => {
+    switch (String.fromCharCode(cp)) {
+      case ' ':
+        return '!_';
+      case '!':
+        return '!|';
+      default:
+        return `!${String.fromCharCode(cp + 0x21)}`;
+    }
+  });
+stringEscapes['^'.charCodeAt(0)] = '_@';
+stringEscapes['_'.charCodeAt(0)] = '__';
+
+const encodeStringWithEscapes = str =>
+  `s${str.replace(/[\0-!^_]/g, ch => stringEscapes[ch.charCodeAt(0)])}`;
+const decodeStringWithEscapes = encoded => {
+  return encoded.slice(1).replace(/([!_])(.|\n)?/g, (esc, prefix, suffix) => {
+    switch (esc) {
+      case '!_':
+        return ' ';
+      case '!|':
+        return '!';
+      case '_@':
+        return '^';
+      case '__':
+        return '_';
+      default: {
+        const ch = /** @type {string} */ (suffix);
+        (prefix === '!' && ch >= '!' && ch <= '@') ||
+          Fail`invalid string escape: ${q(esc)}`;
+        return String.fromCharCode(ch.charCodeAt(0) - 0x21);
+      }
+    }
+  });
+};
+
+const encodeStringWithoutEscapes = str => `s${str}`;
+const decodeStringWithoutEscapes = encoded => encoded.slice(1);
+
+/**
+ * Encodes an array into a sequence of encoded elements, each terminated by a
+ * space (which is part of the escaped range in encoded strings).
+ *
+ * @param {unknown[]} array
+ * @param {(p: Passable) => string} encodePassable
+ * @returns {string}
+ */
+const encodeArrayWithoutEscapes = (array, encodePassable) => {
+  const chars = ['^'];
+  for (const element of array) {
+    const enc = encodePassable(element);
+    chars.push(enc, ' ');
+  }
+  return chars.join('');
+};
+
+/**
+ * @param {string} encoded
+ * @param {(encoded: string) => Passable} decodePassable
+ * @returns {Array}
+ */
+const decodeArrayWithoutEscapes = (encoded, decodePassable) => {
+  encoded.startsWith('^') || Fail`Encoded array expected: ${encoded}`;
+  const tail = encoded.slice(1);
+  const elements = [];
+  let depth = 0;
+  let nextIndex = 0;
+  let currentElementStart = 0;
+  for (const { 0: ch, index: i } of tail.matchAll(/[\^ ]/g)) {
+    const index = /** @type {number} */ (i);
+    if (ch === '^') {
+      // This is the start of a nested array.
+      // TODO: Since the syntax of nested arrays must be validated as part of
+      // decoding the outer one, consider decoding them here into a shared cache
+      // rather than discarding information about their contents until the later
+      // decodePassable.
```

Comment:
Since the point of this PR is to fix the bad complexity measure of my previous `encodePassable` attempt, I think this TODO is needed. Do I understand correctly that this additional TODO, we still have bad time complexity, and have only fixed bad space complexity?

If this is hard, feel free to file an issue and postpone to another PR. But IIUC, it doesn't look hard, in which case I'd prefer it in this PR.

Reply:
I'd like to keep that separate from the encoding changes, because it's a transparent optimization on top of them. But you're right that the difficulty is low, so I can keep a branch and PR on top of this one.

Reply:
Later PR is fine. File an issue so we're sure to get around to it?

Reply:
#1747

Reply:
Thx. Resolving.